pyasic/miners/backends/auradine.py

```diff
@@ -193,6 +195,42 @@ async def send_config(self, config: MinerConfig, user_suffix: str = None) -> Non
         for key in conf.keys():
             await self.web.send_command(command=key, **conf[key])
 
+    async def upgrade_firmware(self, url: str = None, file_path: str = None, version: str = "latest") -> bool:
+        """
+        Upgrade the firmware of the Auradine device.
+
+        Args:
+            url (str): The URL to download the firmware from.
+            file_path (str): The local file path of the firmware to be uploaded.
+            version (str): The version of the firmware to upgrade to.
+
+        Returns:
+            bool: True if the firmware upgrade was successful, False otherwise.
+        """
+        try:
+            logging.info("Starting firmware upgrade process.")
+
+            if not url and not file_path:
+                raise ValueError("Either URL or file path must be provided for firmware upgrade.")
+
+            if url:
+                # Download the firmware file from the URL
+                async with httpx.AsyncClient() as client:
```

Comment:
Use self.web rather than manually using httpx.asyncclient.

Reply:
The way this is done here is correct - https://github.com/UpstreamData/pyasic/pull/157/commits/eefb055a3f2c67f9eecd250a288a63a1dee31142